src/blog/understanding-ecma-part-1.md

```diff
@@ -0,0 +1,232 @@
+
+... where we take a simple function in the spec and try to understand the notation.
+
+## Preface
+
+Even if you know JavaScript, reading its language specification, [ECMAScript Language specification, or the ECMAScript spec for short](https://tc39.es/ecma262/), can be pretty daunting. At least that's how I felt when I started reading it for the first time.
+
+Let's start with a concrete example and walk through the spec to understand it. The following code demonstrates usage of Object.prototype.hasOwnProperty:
+
+```javascript
+let o = {'foo': 1};
+o.hasOwnProperty('foo'); // true
+o.hasOwnProperty('bar'); // false
+```
+
+In the example, _o_ doesn't have a property called _hasOwnProperty_, so we walk up the prototype chain and look for it. We find it in _o_'s prototype, which is _Object.prototype_.
+
+When trying to find out how _Object.prototype.hasOwnProperty_ is defined, I bumped into pseudocode-like descriptions like this:
+
+> Object.prototype.hasOwnProperty(V)
+>
+> When the hasOwnProperty method is called with argument V, the following steps are taken:
+>
+> 1. Let P be **?** ToPropertyKey(V).
+> 2. Let O be **?** ToObject(this value).
+> 3. Return **?** HasOwnProperty(O, P).
+
+[Spec: Object.prototype.hasOwnProperty](https://tc39.es/ecma262#sec-object.prototype.hasownproperty)
+
+and
+
+> HasOwnProperty(O, P)
+>
+>The **abstract operation** HasOwnProperty is used to determine whether an object has an own property with the specified property key. A Boolean value is returned. The operation is called with arguments O and P where O is the object and P is the property key. This abstract operation performs the following steps:
+>
+> 1. **Assert**: Type(O) is Object.
+> 2. **Assert**: IsPropertyKey(P) is true.
+> 3. Let desc be **?** O.**\[\[GetOwnProperty\]\]**(P).
+> 4. If desc is undefined, return false.
+> 5. Return true.
+
+[Spec: HasOwnProperty](https://tc39.es/ecma262#sec-hasownproperty)
+
+But what's an "abstract operation"? What are the things inside \[\[ \]\]? Why is there a "?" in front of a function? What do the asserts mean?
+
+Let's find out!
+
+## ECMAScript Language Types and ECMAScript Specification Types
+
+Let's start with something that looks familiar. The spec uses values such as undefined, true and false, which we already know from JavaScript. They all are **ECMAScript language values**, values of **ECMAScript language types** which the spec also defines.
+
+The spec also uses ECMAScript language values internally, for example, an internal data type might contain a field whose possible values are true and false.
+
+[Spec: ECMAScript language types](https://tc39.es/ecma262/#sec-ecmascript-language-types)
+
+**ECMAScript specification types** are types that occur only in the ECMAScript spec, but not in the JavaScript language, and which might or might not occur as internal types in a JavaScript engine. In this blog post, we'll get to know the ECMAScript specification type _Record_ (and its subtype _Completion Record_).
```

Comment:
```suggestion
In addition to language types and values, the specification also uses types and values occur only in the ECMAScript spec. These are called **ECMAScript specification types**, and a JavaScript engine does not need to (but is free to) implement them. In this blog post, we'll get to know the ECMAScript specification type _Record_ (and its subtype _Completion Record_).
```

Reply:
(Did a slightly modified version of this suggestion.)